Update NATS trigger to use queue subscriptions

phi2039 opened the pull request:
The NATS trigger currently creates a new standard subscription for each processor instance. As a result, every replica of a deployed function receives a notification for every message published to the configured topic (i.e. every message is processed 'n' times). This essentially eliminates the benefits of distributed event handling (with the exception of some level of increased certainty that an event will be handled at least once).

NATS supports an additional round-robin messaging type that allows subscribers to specify a (worker group) queue name along with a topic pattern when subscribing. The queue acts as a single subscriber, and messages are load-balanced across members of the worker group. See [NATS Docs](https://nats.io/documentation/writing_applications/subscribing/)

This PR updates the NATS trigger to enable processor instances associated with the same trigger definition to join the same queue. A queue name may be supplied or generated automatically.

-- Still completing testing, but wanted to get initial feedback re: the need to continue to allow 'broadcast' subscriptions (i.e. trigger all replicas for every message) and naming convention for generated queue names (should be globally unique per trigger definition)

pavius commented:
Looks great :+1: I'm OK with how you auto-generated the queue name and the fact that the broadcast to all replicas is not currently possible. If at some point we hear anyone actually need it, we can add another configuration flag that chooses between broadcast and queue based subscription.

If you feel like it, you can also add support for queue name templating. It's only a few lines of code and will allow users to customize the queuename given namespace, name and trigger id so that the queue names will follow their own convention. If you don't, that's fine.

phi2039 commented:
Love the template idea. Adding that now, then I'll push and remove the WIP.

phi2039 commented:
The template code is in place, but....It turns out that the RuntimeConfiguration.Meta values (Namespace, Name) are not making their way into the factory.Create method. As such, the configured templates (and the default template) cannot use them to generate unique queue names. I see that the RMQ trigger uses these, as well, but I do not have an environment in which to test that right now, so I'm just guessing when I say that the issue is likely occurring there as well.

I'm going to see if I can find some time to dig into why this is happening, but would welcome comments from anyone who has a more full-featured dev/debug setup. I'm going to hold off on finishing the PR until I get this resolved.

pavius commented:
When I check w/the debugger I don't see an issue:
![image](https://user-images.githubusercontent.com/423987/45473762-8f9cad80-b740-11e8-8bd2-9a70ff09b000.png)

If you add a log to Start() with these parameters do you see them empty? Perhaps the issue is with how they are passed to the templating package (I usually copy them to an ad-hoc structure with all the variables at root level)

phi2039 commented:
Thanks @pavius . I am actually doing just that (copying into an ad-hoc map at root level). The map and template code are working properly, as the Id field gets set properly, but when I added a log call to Start(), the Meta values came in empty when running on my k8s cluster (GKE). Since it appears from the above that it is not a completely missing linkage somewhere, I'll take some time to get a more workable dev setup going and diagnose the issue. I updated the PR with the latest code if you're interested. I appreciate all the help thus far.

pavius commented:
Sounds good. I'm curious if a simple log at Start() actually shows that these fields are empty?

phi2039 commented:
It does, oddly enough... This:
![image](https://user-images.githubusercontent.com/859293/45556965-86781300-b80a-11e8-839d-00d9157659a8.png)

Produces this when run locally:
![image](https://user-images.githubusercontent.com/859293/45557033-af000d00-b80a-11e8-9b27-9e6e19aea56c.png)

And this when run on k8s:
![image](https://user-images.githubusercontent.com/859293/45558693-837f2180-b80e-11e8-9d06-00a19035ea3e.png)

pavius commented:
This could be a deficiency when deploying to k8s that never surfaced. We will take a look

pavius commented:
Yes, it seems like it:

https://github.com/nuclio/nuclio/blob/88cba2a351755b4e6812bc7051cab36ffca439be/pkg/platform/kube/functionres/lazy.go#L944

Here we populate the configmap that the function uses and we only populate spec. I think this should be:

```
	if err := configWriter.Write(&configMapContents, &processor.Configuration{
		Config: functionconfig.Config{
			Meta: functionconfig.Meta{
				Name: function.Name,
				Namespace: function.Namespace,
				Labels: labels,
			},
			Spec: function.Spec,
		},
	}); err != nil {
```

We will need to update this and test properly (Sunday). You can try to patch this locally and rebuild the controller... if you're up for that :) Thanks for the heads up!

phi2039 commented:
You are most welcome, and I'm happy to help. I'm glad you were able to track it down so quickly. The abstractions employed are all very elegant, but it can certainly take a bit of time to get familiar enough with them to effective troubleshoot. Thanks for speeding that up!

I can definitely go ahead and patch the controller and deploy it to my GKE cluster to test. I'll circle back once I have that done and dusted.

phi2039 commented:
I have confirmed that this PR is working as expected. Queue auto-naming and templated names will work once #946 is committed.

pavius commented:
Awesome :) I have merged #946. Would you like me to merge this PR as well?

phi2039 commented:
Yes, please :-) This one is good to go. 

Thanks for all the help. Looking forward to continuing to contribute to a great project!

pavius commented:
:+1::+1:

pavius commented:
Oh, note that merging to development auto builds / releases the "unstable" images (takes about 20 minutes). So if you want to use images from upstream this is the way to go until 0.5.9 is released in a few days. Thanks for the great contribution!

phi2039 commented:
@pavius That is GREAT to know, and will save me a lot of trouble. Thanks!

...and you are, of course, most welcome. Thanks for the great platform!
